File: leetcode/Algorithms/662.MaximumWidthOfBinaryTree/Solution.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        int result = 0;
        if (root == null) return result;
        Queue<TreeNode> queue = new LinkedList<>();
        Queue<Integer> index = new LinkedList<>();
        queue.add(root);
        index.add(0);

        while (queue.size() != 0) {
            int size = queue.size(), max = Integer.MIN_VALUE, min = Integer.MAX_VALUE;
            while (size-- > 0) {
                TreeNode tmp = queue.poll();
                int k = index.poll();
                max = Math.max(max, k);
                min = Math.min(min, k);

                if (tmp.left != null) {
                    queue.add(tmp.left);
                    index.add(k * 2 + 1);
                }

                if (tmp.right != null) {
                    queue.add(tmp.right);
                    index.add(k * 2 + 2);
                }
            }
            result = Math.max(result, max - min + 1);
        }
        return result;
    }
}

File: leetcode/Algorithms/662.MaximumWidthOfBinaryTree/Solution.py
class Solution:
    def __init__(self):
        self.left = {}
        self.right = {}

    def widthOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is not None:
            self.helper(root, 0, 0)

        result = 0
        for level in self.left:
            tmp = self.right[level] - self.left[level] + 1
            if result < tmp:
                result = tmp
        return result

    def helper(self, node, level, weight):
        if self.left.get(level, 2147483648) > weight:
            self.left[level] = weight

        if self.right.get(level, -2147483649) < weight:
            self.right[level] = weight

        if node.left is not None:
            self.helper(node.left, level + 1, weight * 2 + 1)

        if node.right is not None:
            self.helper(node.right, level + 1, weight * 2 + 2)

